### ISSUE TITLE
@extending a selector within the same @media directive still gives deprecation warning

### ISSUE THREAD

**Issue by MrOrz** [sass/sass#905]
I know that `@extend` an selector outside an `@media` directive is not supported before the browser supports `@extend` natively.

Therefore I came up with a work-around : Put my class (or placeholder) within a mixin and include it in both inside and outside the `@media` directive. Although it duplicates the included mixin code in the compiled CSS file, it still benefits me because I only need to manage one copy of code in SASS source.

The following code demonstrates the idea explained above:  https://gist.github.com/MrOrz/6373773

``` sass
// A placeholder factory that generates a placeholder.
=btn-placeholder($border-color)
  %btn
    border: 1px solid $border-color
    border-radius: 8px

// Generate a %btn placeholder here.
+btn-placeholder(#000)

// Use the %btn placeholder in the following two classes
.btn-error
  @extend %btn
  background: #f00

.btn-success
  @extend %btn
  background: #0f0

@media screen and (min-width: 750px)

  // Here we invoke the placeholder factory again,
  // this time from within the @media
  +btn-placeholder(#999)

  // The following classes in @media uses the placeholder
  // inside the @media.
  .btn-error
    @extend %btn
    background: #c00

  .btn-success
    @extend %btn
    background: #0c0
```

And it worked great in Sass 3.2.10. The compiled code totally makes sense:

``` css
.btn-error, .btn-success {
  border: 1px solid black;
  border-radius: 8px;
}

.btn-error {
  background: red;
}

.btn-success {
  background: lime;
}

@media screen and (min-width: 750px) {
  .btn-error, .btn-success {
    border: 1px solid #999999;
    border-radius: 8px;
  }

  .btn-error {
    background: #cc0000;
  }

  .btn-success {
    background: #00cc00;
  }
}
```

However, when passed to the compiler it still gives me a deprecation warning. I am worried that this work-around, which is completely valid, would throw an error in Sass 3.3. Please take this work-around into consideration and alter the algorithm of `@extend` error detection.

**Comment by nex3** [comment]
This is a tricky situation. In general, if a selector is defined both inside and outside of a `@media` query and it's `@extend`ed from inside, the user is likely to expect that that `@extend` will work on all instances of the selector. That means it's correct to throw an error in that situation. However, that also means that this workaround that explicitly indicates that the selectors should be duplicated in each context doesn't work.

@chriseppstein, what are your thoughts?

**Comment by xzyfer** [comment]
Although not being able to `@extend` normal selectors within media queries makes perfect sense, IMHO the current behaviour of inserting the placeholder'd content within the media query context that it's being `@extend`ed is intuitive. Losing this feature would really marginalise the general usefulness of placeholders.

**Comment by nex3** [comment]
We certainly don't want placeholders to work differently than normal selectors in any way other than that they don't produce output on their own.

**Comment by xzyfer** [comment]
What's the reasoning behind that though?

You certainly cannot extend normal selectors within a media query because the thing your extending just doesn't existing in that context. However the entire value proposition of placeholders is that they wont exist in a context until extended. There's also the added bonus of being able to write DRY sass when working responsively.

IIRC stylus recommends using placeholders for this exact purpose.

**Comment by nex3** [comment]
The reason we don't allow normal selectors to be extended from within different directives is because we don't want to duplicate the contents of those selectors for each context in which they're used. That duplication would still exist for placeholder selectors if they're extended from multiple contexts.

**Comment by xzyfer** [comment]
I agree that the duplication still needs to happen with placeholders, however I disagree that it's incorrect behaviour. Exactly the opposite in fact. As with @MrOrz I believe the current behaviour of producing output within the context a placeholder was extending is a behaviour that's depended on heavily, and easy to reason about.

In practice if this feature is removed there'll be few to no options for writing clean sass when dealing with media queries. For each context a developer wants to extend a selector they'll either be forced to:
- explicitly duplicate that selector in each context they want to extend it
- wrap that select in a mixin and include it into a each context they want to extend it

Neither of these are intuitive, or clean and the former cannot be DRY. 
IMHO placeholders are conceptually different enough to solve this problem, and easy for developers to reason about.

As a developer, being able to define a placeholder in a single context, and extend it in any context with the knowledge that it will be made available on demand is great. To be honest defining a placeholder in multiple context is counter intuitive because, as you mentioned above, I'd expect all the extended selectors to have been inserted into all context. 

---

If there is fact a way to write DRY sass across multiple media query contexts that I'm not aware of I humbly withdraw my objections here.

**Comment by nex3** [comment]
I'm not sure you completely understand the current behavior. Right now, the following code will result in no output at all:

``` scss
%foo {a: b}

@media screen {.bar {@extend %foo}}
@media print {.baz {@extend %foo}}
```

I don't believe that there is an intuitive way for that example to work. It's unintuitive for the contents of `%foo` to be emitted twice; that's contrary to the way `@extend` works in every other case and risks silently creating a huge amount of unexpected output bloat. It's unintuitive for `%foo` to be silently not extended, because of course the user has indicated that they want it to be extended.

There are several compromises we could have chosen to handle this case. We chose the one we judged most likely to make the user aware of the issue. That way if they want to duplicate the selector they're able to do so explicitly, and if they're able to structure their stylesheet to avoid that they're able to do that as well.

**Comment by xzyfer** [comment]
You're correct about your example, but we can work currently work around that limitation as @MrOrz has shown

``` sass
@mixin foo {
    %foo { a: b; }
}

@include foo;
a { @extend %foo; }
@media screen { @include foo; a { @extend %foo; } }
@media print  { @include foo;  a { @extend %foo; } }
```

I think it make sense for placeholders to be further distinguished from selectors beyond "they don't produce output on their own". Allowing them to be extending into different context's is one such desirable feature. 

The cost of some documentation is small price to pay for the added cleanliness, all around maintainability of large responsive sass projects. 

It's not that I don't understand why you want to remove these features, but I can't help thinking:
- shouldn't developers be making an effort to be aware of the output they're producing?
- does preventing new features, and/or removing existing features, in order to help the lazy/ignorant at the cost of tying the hands of project authors in the best interests of a project? 

---

Personally I would like to see at the very least, is the mixin work around to continue to work, and I'm wishful fo your example to become functional in the same way. Obviously not at the cost of introducing breaking behaviour.

A compromise to achieve the same result, with similar DRY'ness would be to enable @imports within @media. 

``` scss
// component.scss
%foo { a: b; }

// main.scss
@import "component";
a { @extend %foo; }

@media screen { @import "component"; a { @extend %foo; } }
@media print  { @import "component";  a { @extend %foo; } }
```

**Comment by MrOrz** [comment]
I think we can come up with a specific keyword for SASS to copy the specified placeholder from outside the `@media` directory. Something like this should do the trick:

``` sass
%foo
  +link-colors(#aaa, #bbb)

a 
  @extend %foo

@media screen
  @extend %foo !duplicate

@media print
  @extend %foo !duplicate
```

By carefully selecting the keyword that conveys the idea of compiled code bloat, the developers using such keyword would notice the difference between normal `@extend`.

If the approach above seems too automatic, we can also let the developer manually duplicate the placeholder using `@include`, then provide a keyword for them to suppress the deprecation warning or error.  This approach is more flexible because the developer can control how the code is duplicated, like in the example in the beginning of this thread:

``` sass
=btn-placeholder($border-color)
  %btn
    border: 1px solid $border-color
    border-radius: 8px

+btn-placeholder(#000)

.btn-error
  @extend %btn
  background: #f00

@media screen

  +btn-placeholder(#999) // Explicit duplication of %btn done by developer

  .btn-error

    /*
      !use-local is a keyword flag that tells the SASS compiler that
      the developer knows for sure the local copy of %btn is used for @extend.
      Thus the error or warning can be suppressed at ease.

    @extend %btn !use-local
    background: #c00

```

**Comment by jakob-e** [comment]
[Extending within the @media directive](https://github.com/nex3/sass/issues/915)

**Comment by xzyfer** [comment]
@jakob-e I appreciate your work on this. I've had to do some similarly gnarly work arounds myself. Have you ever tried to explain this to new team members? The main point of this thread is that the simple (native) solution to this problem exists and is about to be removed.

@MrOrz I like your general concept. 

What I think could work:
- create some documentation educating developers on some common cases for unintentional output (Nicole Sullivan of OOCSS talks about some of these here https://www.youtube.com/watch?v=GhX8iPcDSsI)
- keep the existing behaviour shown by @MrOrz (for placeholders only IMHO)
- change the warning output from a deprecation warning to `"this could potentially result in duplicate output"` warning with a link to said documentation
- add a directive for suppressing warnings inline e.g. `@extend %placeholder !acknowledge` similar to what exists in [PHP](http://php.net/manual/en/language.operators.errorcontrol.php), [Ruby](http://api.rubyonrails.org/classes/Kernel.html#method-i-suppress), [JShint](https://github.com/jshint/jshint/issues/757#issuecomment-17376443)

I'm suggesting this because it introduces a small number for changes to sass core, as well making an attempt to help move the community as a whole forward with education. It's not new for tools to return useful link and documentation as output. I also think the `"this could potentially result in duplicate output"` warning could be applicable in a number of other use cases in future releases.

Ideally I'd be great if the existing functionality could be extended so that the example demonstrated by @nex3 (https://github.com/nex3/sass/issues/905#issuecomment-23974599) was functional i.e. not needing to wrap placeholders in a mixin inorder to inject them into a context.

**Comment by jakob-e** [comment]
Thanks xzyfer,

What initiated my work was this [question](https://github.com/nex3/sass/issues/640) and the answer provided by [robwierzbowski](https://github.com/nex3/sass/issues/640#issuecomment-12770202) and [nex3](https://github.com/nex3/sass/issues/640#issuecomment-13023139). I see the point in SCSS extends being about moving content and mixins to create/duplicate it.

I don't see any confusion or problems in my [solution](https://github.com/nex3/sass/issues/915) - so I hope you are not telling me that @extends can't be produced by mixins (or passed by @content)... In this case it would cut the frameworkability of SCSS in half :(  
As I see it there are no conflicts in any of the steps below - nor no confusion about what is being extended
...as you only extend from within each media directive.

Am I missing something? 

``` SCSS

// Valid construction
@mixin  mediaExtend(){   // Mixins can create/replicate input and @content
    @for .... through {  // Loop thru each @media directive eg. defined in a list
       @media .... {     // Print out each @media directive
             @content;   // Adding @content (not a reference)
       }
    }
}

// Valid include
@include mediaExtend(){
      %test { content: 'test'; } // @content - same as .test { content:'test'; }
}

// Valid mixin output 
@media all { %test { content: 'test'; } }
@media screen and (min-width: 769px) { %test { content: 'test'; } }
@media screen and (min-width: 481px) and (max-width: 768px) { %test { content: 'test'; } }
@media screen and (max-width: 480px) { %test { content: 'test'; } }

// Valid mixin extend - extending within the directive
@media all { 
    .foo { @extend %test;  } 
}

```

Best,
Jakob E

**Comment by xzyfer** [comment]
You're approach is completely valid. There are many similar approaches around that work on achieving the same thing. 

My point simply is that it's a decent amount of boilerplate for something that could easily, and justifiably exist natively. And in a round about way already does :)

Specifically what I was saying before was trying to explain these kinds of work arounds to new developers that might not be familiar with the quirks in sass that make them necessary is a decent amount of work, and can be barrier to entry. 

The other reason I'm personally speaking out on this topic is because the feature being removed, and amendments being suggested by myself are the recommended workflow in other preprocessors :)

All in all I think there's worth in pushing the language forward, and that sometimes as in this case, means balancing the complexity; practicality; and developer friendliness.

**Comment by jakob-e** [comment]
I'm a newbee to SCSS and find it a bit like the old AS2  - it has a lot of potential but also the flaws of an emerging language that would benefit from a complete hurlover. I don't mind having to do the "classes" myself but as you say there are a lot of quirking involved and the base toolbox still needs a lot of tools like string handling (split/join) list handling (like javascript arrays) and objects of course.

That said I find the entry level extremely simple - just do CSS and then add on as you get hooked by the wonderful nesting and extends :)

On the topic of removing features to accommodate ignorants - I agree 100% 
We are responsible - let us mess up and learn :)

How about adding a turn off warnings feature ?

**Comment by xzyfer** [comment]
If they were to go down the warning suppression route, I'd opt for doing so inline via a language directive.
That way you're explicitly acknowledging you understand what's happening in this single case. 
Blanketly suppressing errors can only do harm IMO.

**Comment by jakob-e** [comment]
If SCSS is to be a thing for the masses my guess is that most users won't use ruby - but a tool like **[CodeKit](http://incident57.com/codekit/)**. If the compiler is not able to handle warnings you are worse off with them than without. Officially this is not a SCSS problem but nor is it a CSS problem the way IE8 handles things - if you get my point ;) 

**Example** 
Try this using ruby (quiet all deprecation warnings)
... and then try it using CodeKit. 

``` SCSS

.extend { content:'A'}
@media all {.extend { content:'B'} }

@mixin extendTest($times){
    @for $i from 0 through $times{
        .test#{$i} { @extend .extend; }
        @media all { .test#{$i} { @extend .extend; } }
    } 
}

// First try this
@include extendTest(20); // Works and output is perfectly valid

// And then this 
// @include extendTest(50); // Nothing, no warnings, just nothing 

// My guess is that the warnings makes CodeKit choke
// - but that is just a big maybe


```

Once again - I can't get my head around why this wonderful feature is being deprecated :(

Best,
Jakob E

**Comment by jakob-e** [comment]
And by the way (if possible) being able to set options from within SCSS would be nice 

Like:
@option style:nested;
@option cache:true;
@option quiet:true;

**Comment by jakob-e** [comment]
Just added an update to my [solution](https://github.com/nex3/sass/issues/915) to show how much complexity is added by deprecating this feature - It seems to do the trick of having only one source to maintain\* - but boy do we have to travel this road?  
- fingers crossed hope this does not throw warnings

**Comment by chriseppstein** [comment]
Closing this issue. It's not a bug. It's the way Sass is going to work.

Between the `unique-id()` helper and `@at-root`, it will not be that hard to support different medias.

I will be blogging on how to work around the new limitation using these features.

**Comment by xzyfer** [comment]
@chriseppstein do have an initial draft on this, or could you point me in the direction to investigate myself?

I could be mistaken but I've yet to see this addressed either in the spec outlined in #774 or the test cases added for the resulting `@at-root` pull requests.

> Specifically, this allows you to extend a selector via a mixin for all media contexts instead of just within the current media block.

I'm referring specifically to the use case of defining a `%placeholder` or `.normal-class` once without any `@media` directives and being able to **extend a [it] via a mixin for all media contexts instead of just within the current media block**
